src/passes/function_passes/LiveVariablesPass.cpp

```diff
@@ -0,0 +1,153 @@
+#include <vector>
+#include <stdint.h>
+#include <llvm/Pass.h>
+#include <llvm/IR/Intrinsics.h>
+#include <llvm/IR/IRBuilder.h>
+#include <llvm/IR/BasicBlock.h>
+#include <llvm/IR/Function.h>
+#include <llvm/IR/Module.h>
+#include <llvm/IR/DataLayout.h>
+#include <llvm/IR/Instruction.h>
+#include <llvm/IR/LegacyPassManager.h>
+#include <llvm/IR/Constants.h>
+#include <llvm/IR/Type.h>
+#include <llvm/IR/Value.h>
+#include <llvm/Support/raw_ostream.h>
+#include <llvm/Transforms/IPO/PassManagerBuilder.h>
+#include <llvm/IR/Dominators.h>
+#include <llvm/Transforms/Utils/BasicBlockUtils.h>
+
+#define UNOPT_PREFIX "__unopt_"
+
+using namespace llvm;
+using std::vector;
+
+namespace {
+
+/*
+ * Searches for each call to `llvm.experimental.stackmap` or
+ * `llvm.experimental.patchpoint` and passes the variables which are live at
+ * the callsite as arguments to the call.
+ */
+struct LiveVariablesPass: public FunctionPass {
+  static char id;
+
+  LiveVariablesPass() : FunctionPass(id) {}
+
+  virtual bool runOnFunction(Function &fun) {
+    Module *mod = fun.getParent();
+    StringRef funName = fun.getName();
+    outs() << "Running LiveVariablesPass on function: " << funName << '\n';
+    for (auto &bb : fun) {
+      for (BasicBlock::iterator it = bb.begin(); it != bb.end(); ++it) {
+        if (isa<CallInst>(it)) {
+          CallInst &callInst = cast<CallInst>(*it);
+          Function *calledFun = callInst.getCalledFunction();
+          if (callInst.isInlineAsm()) {
+            continue;
+          }
+          if (producesStackmapRecords(calledFun)) {
+            // This is a stackmap/patchpoint call, so it needs to record all
+            // the variables live at this point.
+            auto liveVariables = getLiveRegisters(*it);
+            vector<Value *> args(callInst.arg_begin(), callInst.arg_end());
+            // Pass the live locations to the stackmap/patchpoint call. This
+            // also inserts the size of each 'live' location into the stackmap.
+            std::move(liveVariables.begin(), liveVariables.end(),
+                      std::back_inserter(args));
+            CallInst *newCall = CallInst::Create(calledFun, args);
+            if (!callInst.use_empty()) {
+              callInst.replaceAllUsesWith(newCall);
```

Comment:
OK, so this code just mutates the arguments to your stackmap calls to add the live vars?

Reply:
Yep, that's pretty much it.